lib/SILOptimizer/LoopTransforms/COWArrayOpt.cpp

```diff
@@ -1446,21 +933,102 @@ bool COWArrayOpt::hasLoopOnlyDestructorSafeArrayOperations() {
   return ReturnWithCleanup(true);
 }
 
-/// Hoist the make_mutable call and optionally the projection chain that feeds
-/// the array self argument.
-void COWArrayOpt::hoistMakeMutableAndSelfProjection(
-    ArraySemanticsCall MakeMutable, bool HoistProjection) {
-  // Hoist projections.
-  if (HoistProjection)
-    hoistAddressProjections(MakeMutable.getSelfOperand(),
-      Preheader->getTerminator(), DomTree);
+/// Return the underlying Array address after stripping off all address
+/// projections. Returns an invalid SILValue if the array base does not dominate
+/// the loop.
```

Comment:
It's especially important for 2d arrays. But it can an appear whenever the array is contained in another struct/class.
There are test cases in cowarray_opt.sil for this.